### PR TITLE
fix linker flags in cmake (find_package based) generators

### PR DESCRIPTION
**Pull request by memsharded**
Changelog: Bugfix: Fix linker flags in cmake (find_package based) generators.
Docs: Omit

Fix: https://github.com/conan-io/conan/issues/8703

#tags: slow

### CODE REVIEW

File: conan/tools/cmake/cmakedeps.py [hunk 1]
```diff
@@ -235,8 +235,9 @@ def join_paths_single_var(values):
         # Issue: #1251
         self.cxxflags_list = join_flags(";", cpp_info.cxxflags)
         self.cflags_list = join_flags(";", cpp_info.cflags)
-        self.sharedlinkflags_list = join_flags(";", cpp_info.sharedlinkflags)
-        self.exelinkflags_list = join_flags(";", cpp_info.exelinkflags)
+        self.sharedlinkflags_list = join_flags(";", ["-{}".format(f)
+                                                     for f in cpp_info.sharedlinkflags])
+        self.exelinkflags_list = join_flags(";", ["-{}".format(f) for f in cpp_info.exelinkflags])
```
Comment (srnwk) [inline conan/tools/cmake/cmakedeps.py]: Could this break flags that already start with a `-` or `--`, for non-MSVC linkers?

Comment (memsharded) [reply conan/tools/cmake/cmakedeps.py]: Yes, good catch, I will verify that too.

Comment (memsharded) [reply conan/tools/cmake/cmakedeps.py]: Keeping flags that already start with "-" in https://github.com/conan-io/conan/pull/8740/commits/4ebe404bbd85c43c71a44626816dd419dc37c6b1

File: conans/test/unittests/client/generators/cmake_find_package_multi_test.py [hunk 2]
```diff
@@ -27,3 +28,27 @@ def test_cmake_find_package_multi_version(self):
...
+    cpp_info = CppInfo("mypkg", "dummy_root_folder1")
+    cpp_info.sharedlinkflags = ["/NODEFAULTLIB", "/OTHERFLAG"]
+    cpp_info.exelinkflags = ["/OPT:NOICF"]
+    conanfile.deps_cpp_info.add("mypkg", cpp_info)
+
+    gen = CMakeFindPackageMultiGenerator(conanfile)
+    files = gen.content
+    d = files["mypkgTarget-release.cmake"]
+    assert "$<$<STREQUAL:$<TARGET_PROPERTY:TYPE>,EXECUTABLE>:-/OPT:NOICF>" in d
+    assert "$<$<STREQUAL:$<TARGET_PROPERTY:TYPE>,MODULE_LIBRARY>:-/NODEFAULTLIB;-/OTHERFLAG>" in d
```
Comment (SSE4) [inline conans/test/unittests/client/generators/cmake_find_package_multi_test.py]: so, is it really expected to pass `-/OPT` with both dash and slash?

Comment (memsharded) [reply conans/test/unittests/client/generators/cmake_find_package_multi_test.py]: Apparently it worked, but you are right, better double check, I'll check it.

File: conan/tools/cmake/cmakedeps.py [hunk 3]
```diff
@@ -216,6 +216,9 @@ def join_paths_single_var(values):
             """
             return '"%s"' % ";".join(p.replace('\\', '/').replace('$', '\\$') for p in values)
 
+        def format_link_flags(link_flags):
+            return [(f if f.startswith("-") else"-{}".format(f)) for f in link_flags]
```
Comment (danimtb) [inline conan/tools/cmake/cmakedeps.py]:
```suggestion
            return [(f if f.startswith("-") else "-{}".format(f)) for f in link_flags]
```

File: conans/client/generators/cmake.py [hunk 4]
```diff
@@ -34,6 +34,9 @@ def join_paths_single_var(values):
             """
             return '"%s"' % ";".join(p.replace('\\', '/').replace('$', '\\$') for p in values)
 
+        def format_link_flags(link_flags):
+            return [(f if f.startswith("-") else"-{}".format(f)) for f in link_flags]
```
Comment (danimtb) [inline conans/client/generators/cmake.py]:
```suggestion
            return [(f if f.startswith("-") else "-{}".format(f)) for f in link_flags]
```

File: conans/test/unittests/client/generators/cmake_find_package_multi_test.py (hunk 2 above)
Comment (danimtb) [reply conans/test/unittests/client/generators/cmake_find_package_multi_test.py]: yeah, that syntax looks a bit weird to me

Comment (memsharded) [reply conans/test/unittests/client/generators/cmake_find_package_multi_test.py]: Ok, it was working, it didn't raise an error, but just a warning. Dropped the extra /, please check.

**Review by danimtb — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Pull request merged**